# 02 — Feature Engineering
### NexaTel Customer Churn Prediction Project

**Goal:** turn raw columns into signals a model can actually learn from, fix the data-quality issues found in EDA, and document *why* each new feature should help — before we ever train a model.

All feature-engineering logic lives in `ml/feature_engineering.py` as a shared module, so the exact same code runs here, in preprocessing, and in the live backend at prediction time (no train/serve skew).


In [1]:
import os, sys
sys.path.insert(0, '..')
import pandas as pd
from sqlalchemy import create_engine
from dotenv import load_dotenv
from ml.feature_engineering import clean_raw, engineer_features, FEATURE_JUSTIFICATIONS

load_dotenv()
DATABASE_URL = os.getenv("DATABASE_URL", "postgresql://nexatel_user:nexatel_dev_pw@localhost:5432/nexatel_churn")
engine = create_engine(DATABASE_URL)

df = pd.read_sql("SELECT * FROM v_customer_360;", engine)
print(df.shape)

(7043, 21)


## 1. Fix issues found in EDA

In [2]:
print("Missing total_charges before cleaning:", df['total_charges'].isna().sum())
df_clean = clean_raw(df)
print("Missing total_charges after cleaning:", df_clean['total_charges'].isna().sum())
df_clean.loc[df['tenure'] == 0, ['customer_id', 'tenure', 'monthly_charges', 'total_charges']].head()

Missing total_charges before cleaning: 11
Missing total_charges after cleaning: 0


,customer_id,tenure,monthly_charges,total_charges
488,4472-LVYGI,0,52.55,52.55
753,3115-CZMZD,0,20.25,20.25
937,5709-LVOEQ,0,80.85,80.85
1082,4367-NUYAO,0,25.75,25.75
1340,1371-DWPAZ,0,56.05,56.05


## 2. Engineer new features

In [3]:
df_fe = engineer_features(df_clean)
new_cols = ['tenure_group', 'total_services', 'avg_monthly_spend_ratio',
            'high_risk_flag', 'payment_risk_flag', 'contract_ordinal']
df_fe[['customer_id'] + new_cols].head(10)

,customer_id,tenure_group,total_services,avg_monthly_spend_ratio,high_risk_flag,payment_risk_flag,contract_ordinal
0,7590-VHVEG,0-12,1,29.85,1,1,0
1,5575-GNVDE,25-48,2,55.57,0,1,1
2,3668-QPYBK,0-12,2,54.08,1,1,0
3,7795-CFOCW,25-48,3,40.91,0,0,1
4,9237-HQITU,0-12,0,75.82,1,1,0
5,9305-CDSKC,0-12,3,102.56,0,1,0
6,1452-KIOVK,13-24,2,88.61,0,0,0
7,6713-OKOMC,0-12,1,30.19,0,1,0
8,7892-POOKP,25-48,4,108.79,0,1,0
9,6388-TABGU,49+,2,56.26,0,0,1


## 3. Feature Justification Table

In [4]:
import pandas as pd
justification_table = pd.DataFrame([
    {"feature": k, "justification": v} for k, v in FEATURE_JUSTIFICATIONS.items()
])
pd.set_option('display.max_colwidth', None)
justification_table

,feature,justification
0,tenure_group,"Bucketing raw tenure (0-12, 13-24, 25-48, 49+ months) lets tree models split on a coarser, more stable signal and lets the linear model treat 'new customer' as a discrete risk category. EDA (Phase 2) showed churn risk is heavily concentrated in the first 12 months, then drops off — a non-linear relationship a single continuous tenure coefficient can't fully capture."
1,total_services,"Counts how many of the 6 add-on services (security, backup, device protection, tech support, streaming TV, streaming movies) a customer has. SQL/EDA findings show churn falls steadily as this count rises (from ~46% at 1 service down to ~5% at 6) — this single number compresses six categorical columns into one strong, monotonic-ish signal."
2,avg_monthly_spend_ratio,"TotalCharges / tenure, i.e. a customer's true average spend per month over their whole relationship (vs. their *current* MonthlyCharges, which may have changed). Large gaps between this and current MonthlyCharges can indicate a recent plan change, which is itself often a precursor to churn. For tenure = 0 customers, this is set equal to MonthlyCharges (their only observed month) rather than left undefined."
3,high_risk_flag,A binary flag for the exact segment SQL analysis found to be sharpest: tenure < 6 months AND contract = Month-to-month AND no tech support. That segment churns at 66.5% vs. 26.5% overall. Giving the model this combination directly (not just the three underlying columns) makes it easier for linear models to pick up an interaction that in reality is highly non-additive.
4,payment_risk_flag,"A binary flag for manual payment methods (Electronic check or Mailed check) vs. automatic payment (Bank transfer or Credit card, both automatic). SQL findings show manual/electronic-check payers churn at roughly 3x the rate of autopay customers — likely because manual payment creates a natural monthly 'should I keep this?' decision point autopay customers never face."
5,contract_ordinal,"Contract has a genuine order (Month-to-month < One year < Two year) in terms of commitment level, so it is ordinal-encoded (0/1/2) rather than one-hot encoded for models where that ordering is meaningful signal on its own (and one-hot encoded in parallel for models that don't assume order)."


## 4. Sanity-check the new features against churn

In [5]:
print("Churn rate by high_risk_flag:")
print(df_fe.groupby('high_risk_flag')['churn'].mean().round(4) * 100)
print()
print("Churn rate by payment_risk_flag:")
print(df_fe.groupby('payment_risk_flag')['churn'].mean().round(4) * 100)
print()
print("Churn rate by total_services:")
print(df_fe.groupby('total_services')['churn'].mean().round(4) * 100)

Churn rate by high_risk_flag:


high_risk_flag
0    20.62
1    66.70
Name: churn, dtype: float64

Churn rate by payment_risk_flag:
payment_risk_flag
0    15.98
1    34.67
Name: churn, dtype: float64

Churn rate by total_services:
total_services
0    21.41
1    45.76
2    35.82
3    27.37
4    22.30
5    12.43
6     5.28
Name: churn, dtype: float64


**Confirmed:** `high_risk_flag` isolates a 66.5% churn segment (matches the SQL finding exactly, since it's built from the same three columns). `payment_risk_flag` and `total_services` both show a clear, monotonic-ish relationship with churn, confirming they carry real signal before we've trained anything.

## 5. Data Leakage Check

Every engineered feature was checked against one rule: **could this value exist, unchanged, on the day *before* the customer decided whether to churn?**

| Feature | Built from | Leakage risk? |
|---|---|---|
| `tenure_group` | `tenure` (known at any point in the relationship) | None |
| `total_services` | current service subscriptions | None — these are current state, not post-cancellation state |
| `avg_monthly_spend_ratio` | `total_charges` / `tenure` (both known pre-decision) | None |
| `high_risk_flag` | `tenure`, `contract`, `tech_support` (all known pre-decision) | None |
| `payment_risk_flag` | `payment_method` (known pre-decision) | None |
| `contract_ordinal` | `contract` (known pre-decision) | None |

No engineered feature uses `churn` itself, a churn date, a cancellation reason, or any field that would only be populated *after* a customer has already left. This matters because `total_charges` and `tenure` do co-move with churn (shorter relationships = smaller total_charges), which could look like leakage at a glance — but both are legitimately known at prediction time, before any churn decision, so they are safe to use.

In [6]:
df_fe.to_csv('../data/processed/features_engineered.csv', index=False)
print("Saved engineered feature set:", df_fe.shape)

Saved engineered feature set: (7043, 27)
